### PR TITLE
build(docs-infra): several updates/improvements for the PR preview server

### PR DESCRIPTION
**Pull request by gkalpak**
See individual commits for details.

### CODE REVIEW

**Comment by petebacondarwin** [comment]
Tiny typo in the 3rd commit: `This helps surface such error.` -> `This helps surface such errors.`

**Comment by gkalpak** [comment]
Thx, fixed the typo.

File: aio/aio-builds-setup/dockerbuild/scripts-sh/health-check.sh [hunk 1]
```diff
@@ -8,10 +8,32 @@ exitCode=0
...
 # Helpers
+function checkCert {
+  local certPath=$1
+
+  if [[ ! -f "$certPath" ]]; then
+    echo "Certificate '$certPath' does not exist. Skipping expiration check..."
+    return
+  fi
+
+  openssl x509 -checkend 0 -in "$certPath" -noout > /dev/null
```
Comment (petebacondarwin) [inline aio/aio-builds-setup/dockerbuild/scripts-sh/health-check.sh:19]: As I understand it `-checkend X` is checking whether the cert expires in the next X seconds. If so, is it worth giving us some notice, like setting X to something larger like 1 hour?

File: aio/aio-builds-setup/dockerbuild/Dockerfile [hunk 2]
```diff
@@ -79,13 +79,18 @@ RUN echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources
 
 
 # Install packages
+# NOTE: Some packages (such as `nginx`, `nodejs`, `openssl`) make older versions unavailable on the
+#       repositories, so we cannot pin to specific versions for these packages :(
```
Comment (petebacondarwin) [inline aio/aio-builds-setup/dockerbuild/Dockerfile:83]: Strictly we can pin them but we would need to update them in a timely fashion. By not pinning we are relaxing the versions in order to avoid unwanted outages where we have not updated the versions soon enough. Right?

File: aio/aio-builds-setup/scripts/update-preview-server.sh [hunk 3]
```diff
@@ -7,9 +7,9 @@ echo -e "\n\n[`date`] - Updating the preview server..."
 
 # Input
 readonly HOST_REPO_DIR=$1
-readonly HOST_LOCALCERTS_DIR=$2
-readonly HOST_SECRETS_DIR=$3
-readonly HOST_BUILDS_DIR=$4
+readonly HOST_SECRETS_DIR=$2
+readonly HOST_BUILDS_DIR=$3
+readonly HOST_LOCALCERTS_DIR=$4
```
Comment (petebacondarwin) [inline aio/aio-builds-setup/scripts/update-preview-server.sh:12]: What? Not alphabetical??

**Review by petebacondarwin — APPROVED** [review]
A few nits. Nice clean up

**Comment by AndrewKushnir** [comment]
@gkalpak FYI I've removed "merge" label for now and added "cleanup", since there are some comments from Pete. Please put the "merge" label back when this PR is ready to be merged. Thank you.

**Comment by petebacondarwin** [comment]
Oops! Sorry I forgot to add the cleanup label.

File: aio/aio-builds-setup/dockerbuild/scripts-js/test/common/github-pull-requests.spec.ts [hunk 4]
```diff
@@ -47,7 +47,7 @@ describe('GithubPullRequests', () => {
 
 
     it('should make a POST request to Github with the correct pathname, params and data', () => {
-      githubApi.post.and.callFake(() => Promise.resolve());
+      githubApi.post.and.callFake(() => Promise.resolve() as Promise<any>);
```
[the remark this answers is not in the document]
Comment (gkalpak) [reply aio/aio-builds-setup/dockerbuild/scripts-js/test/common/github-pull-requests.spec.ts:50]: Oh, I didn't think of it (latest JAsmine is new to me :grin:).
I'll give it a go. Thx!

File: aio/aio-builds-setup/dockerbuild/scripts-sh/health-check.sh (hunk 1 above)
Comment (gkalpak) [reply aio/aio-builds-setup/dockerbuild/scripts-sh/health-check.sh:19]: This is mainly for detecting outdated, locally generated, self-signed certs. Theoretically, they are generated on every `docker build`, but Docker might use an intermediate image layer from the cache and not run the command.

And on production, the health-check is not the best way to get notified about soon-to-expire certs :grin:
Not to mention that we are currently not monitoring the health-check result in real-time.

Bottom line: I don't think this would provide any benefit as things are set up currently.

File: aio/aio-builds-setup/dockerbuild/Dockerfile (hunk 2 above)
Comment (gkalpak) [reply aio/aio-builds-setup/dockerbuild/Dockerfile:83]: Right.

File: aio/aio-builds-setup/scripts/update-preview-server.sh (hunk 3 above)
Comment (gkalpak) [reply aio/aio-builds-setup/scripts/update-preview-server.sh:12]: I wondered about that too. It's very not-like-me to have these in another order, right? 🤔
I guess one reason is that [the last three][1] are optional, so it wouldn't make sense to have them interleaved with the mandatory ones.

[1]: https://github.com/angular/angular/blame/1d9375d115b1799cd7335a0b410406b84c8ee4cb/aio/aio-builds-setup/docs/vm-setup--start-docker-container.md#L58-L71

File: aio/aio-builds-setup/dockerbuild/scripts-js/test/common/github-pull-requests.spec.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (gkalpak) [reply aio/aio-builds-setup/dockerbuild/scripts-js/test/common/github-pull-requests.spec.ts]: Added a fixup commit. Worked like a charm 🤩

**Comment by petebacondarwin** [comment]
There are a few more places that could use `resolveTo()` and also there are some places that could use `rejectTo()`...

**Comment by gkalpak** [comment]
Added 5 more commits (at the end), mainly "modernizing" all preview server tests (and some other small fixes):
- **build(docs-infra): remove linting from the preview server `dev` npm script** (dd69303d424fcaefde32455fa55bd4f347e2a29b)
- **fix(docs-infra): exit with an error when cleaning up the preview server fails** (101a2b242e97997453d10a80c57241d48d2bc18a)
- **test(docs-infra): remove incorrect preview server test** (6814f7aca186cc4289597e965f66f8c43c297b88)
- **test(docs-infra): enable accidentally disabled preview server tests** (f928bde1d04e65b64145f5b1c4a3994e784af398)
- **refactor(docs-infra): take advantage of latest Jasmine features in preview server tests** (845036a7d807781336d6c760091751ece5ae2fea)

@petebacondarwin, ptal 😃

File: aio/aio-builds-setup/dockerbuild/scripts-js/test/preview-server/build-creator.spec.ts [hunk 5]
```diff
@@ -67,178 +69,152 @@ describe('BuildCreator', () => {
...
+      it('should return a promise', async () => {
         const promise = bc.create(pr, sha, archive, isPublic);
-        promise.then(done);   // Do not complete the test (and release the spies) synchronously
-                              // to avoid running the actual `extractArchive()`.
+        expect(promise).toBeInstanceOf(Promise);
 
-        expect(promise).toEqual(jasmine.any(Promise));
+        // Do not complete the test (and release the spies) synchronously to avoid running the actual
+        // `extractArchive()`.
+        await promise;
```
Comment (petebacondarwin) [inline aio/aio-builds-setup/dockerbuild/scripts-js/test/preview-server/build-creator.spec.ts:78]: If we don't care for the promise to complete then perhaps this test should not be `async`? E.g. 

```ts
it('should return a promise', () => {
  const promise = bc.create(pr, sha, archive, isPublic);
  expect(promise).toBeInstanceOf(Promise);
});
```

File: aio/aio-builds-setup/dockerbuild/scripts-js/test/preview-server/build-creator.spec.ts [hunk 6]
```diff
@@ -517,78 +480,68 @@ describe('BuildCreator', () => {
...
-          done();
-        });
 
-        shellRmSpy.and.callFake(() => {
-          // tslint:disable-next-line: no-string-throw
-          throw 'Test';
-        });
+      it('should abort and reject if it fails to remove the build artifact file', async () => {
+        // tslint:disable-next-line: no-string-throw
+        shellRmSpy.and.callFake(() => { throw 'Test'; });
```
Comment (petebacondarwin) [inline aio/aio-builds-setup/dockerbuild/scripts-js/test/preview-server/build-creator.spec.ts:537]: Could this be?

```suggestion
        shellRmSpy.and.throwError('Test');
```

File: aio/aio-builds-setup/dockerbuild/scripts-js/test/preview-server/build-creator.spec.ts [hunk 7]
```diff
@@ -609,85 +562,72 @@ describe('BuildCreator', () => {
...
-    it('should return a promise', done => {
+    it('should return a promise', async () => {
       const promise = (bc as any).listShasByDate('input/dir');
-      promise.then(done);   // Do not complete the test (and release the spies) synchronously
-                            // to avoid running the actual `ls()`.
+      expect(promise).toBeInstanceOf(Promise);
 
-      expect(promise).toEqual(jasmine.any(Promise));
+      // Do not complete the test (and release the spies) synchronously to avoid running the actual `ls()`.
+      await promise;
```
Comment (petebacondarwin) [inline aio/aio-builds-setup/dockerbuild/scripts-js/test/preview-server/build-creator.spec.ts:570]: See other similar test comment.

**Review by petebacondarwin — APPROVED** [review]

File: aio/aio-builds-setup/dockerbuild/scripts-js/test/common/github-teams.spec.ts [hunk 8]
```diff
@@ -16,6 +16,7 @@ describe('GithubTeams', () => {
       expect(() => new GithubTeams(githubApi, '')).
         toThrowError('Missing or empty required parameter \'githubOrg\'!');
     });
+
```
[the remark this answers is not in the document]
Comment (gkalpak) [reply aio/aio-builds-setup/dockerbuild/scripts-js/test/common/github-teams.spec.ts:19]: Consistency :grin:

File: aio/aio-builds-setup/dockerbuild/scripts-js/test/preview-server/build-creator.spec.ts (hunk 5 above)
Comment (gkalpak) [reply aio/aio-builds-setup/dockerbuild/scripts-js/test/preview-server/build-creator.spec.ts:78]: As mentioned in the comment, we should _not_ let the test complete before the promise completes, because what would release the spies and we might interact with actual implementations (which is not desirable).

So, while we don't indeed care about the resolved value, we need to wait out the promise.

File: aio/aio-builds-setup/dockerbuild/scripts-js/test/preview-server/build-creator.spec.ts (hunk 6 above)
Comment (gkalpak) [reply aio/aio-builds-setup/dockerbuild/scripts-js/test/preview-server/build-creator.spec.ts:537]: It could, but that is slightly different. `.and.throwError('Test')` is equivalent to `.and.callFake(() => { throw new Error('Test'); })`.

File: aio/aio-builds-setup/dockerbuild/scripts-js/test/preview-server/build-creator.spec.ts (hunk 7 above)
Comment (gkalpak) [reply aio/aio-builds-setup/dockerbuild/scripts-js/test/preview-server/build-creator.spec.ts:570]: See https://github.com/angular/angular/pull/36837#discussion_r419525399.
